File: day3.py
with open("day3.txt", 'r') as f:
    bits = [list(line.strip()) for line in f]

gamma = ""
epsilon = ""

for index, _ in enumerate(bits[0]):
    count = sum(line[index] == '1' for line in bits)
    state = bool(count > len(bits) / 2)
    #print(index, count, state)
    gamma += '1' if state else '0'
    epsilon += '0' if state else '1'



print(f"Solution part 1: { int(gamma, 2) * int(epsilon, 2) }")

# create a copy of bits to work with
com = bits[::]
for index in range(len(com[0])):
    if len(com) == 1:
        num = com[:index+1]
        break
    else:
        common = '1' if sum(line[index] == '1' for line in com) >= len(com) / 2 else '0'
        com = [ line for line in com if line[index] == common ]

oxygen = int("".join(com[0]),2)

com = bits[::]
for index in range(len(com[0])):
    if len(com) == 1:
        num = com[:index+1]
        break
    else:
        common = '1' if sum(line[index] == '1' for line in com) >= len(com) / 2 else '0'
        com = [ line for line in com if line[index] != common ]

co2 = int("".join(com[0]),2)

print(f"Solution part 2: { oxygen * co2 }\n{ oxygen }, { co2 }")

File: day3.rs
use std::{
    fs::File,
    io::prelude::*,
    string::*,
};

fn main(){
    let mut file = File::open("day3.txt").expect("file not found");
    let mut buf = String::new();
    file.read_to_string(&mut buf).expect("something went wrong reading the file");
    part1(&buf);

}


fn part1(buf: &str){
    let mut gamma = "".to_string();
    let mut epsilon = "".to_string();

    let mut count;
    for line in buf.lines(){
        count = 0;
        for c in line.chars(){
            if c == '1'{
                count += 1;
            }
        }        
        if count > line.chars().count()/2{
            gamma.push('1');
            epsilon.push('0');
        }
        else{
            gamma.push('0');
            epsilon.push('1');
        }
    }

    println!("part 1 ans: {},{}", gamma, epsilon);

    let intval = u256::from_str_radix("000100110011100101", 2).unwrap();
    println!("{}", intval);


    // let gamma_dec = usize::from_str_radix(gamma.as_str(), 2).unwrap();
    // let epsilon_dec = usize::from_str_radix(epsilon.as_str(), 2).unwrap();
    // getting an overflow on these expressions
   
    // println!("part 1 ans: {}", gamma_dec * epsilon_dec);
}